detection_rules/endgame.py

```diff
@@ -0,0 +1,103 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+
+"""Endgame Schemas management."""
+import json
+import shutil
+import sys
+from pathlib import Path
+
+import eql
+
+from .utils import ETC_DIR, DateTimeEncoder, cached, gzip_compress, read_gzip
+
+ENDGAME_SCHEMA_DIR = Path(ETC_DIR) / "endgame_schemas"
+
+
+class EndgameSchemaManager:
+    """Endgame Class to download, convert, and save endgame schemas from TBD."""
+
+    def __init__(self, github_client, endgame_version: str):
+        self.repo = github_client.get_repo("elastic/endgame-evecs")
+        self.endgame_version = endgame_version
+        self.endgame_schema = self.download_endgame_schema()
+
+    def download_endgame_schema(self) -> dict:
+        """Download schema from endgame-evecs."""
+
+        # Use the static mapping.json file downloaded from the endgame-evecs repo.
+        main_branch = self.repo.get_branch("master")
+        main_branch_sha = main_branch.commit.sha
+        schema_path = "pkg/mapper/ecs/schema.json"
+        endgame_mapping = self.repo.get_contents(schema_path, ref=main_branch_sha)
+        endgame_mapping = json.loads(endgame_mapping.decoded_content.decode())
+
+        return endgame_mapping
+
+    def save_schemas(self, overwrite: bool = False):
+
+        schemas_dir = ENDGAME_SCHEMA_DIR / self.endgame_version
+        if schemas_dir.exists() and not overwrite:
+            raise FileExistsError(f"{schemas_dir} exists, use overwrite to force")
+        else:
+            shutil.rmtree(str(schemas_dir.resolve()), ignore_errors=True)
+            schemas_dir.mkdir()
+
+        # write the raw schema to disk
+        raw_os_schema = self.endgame_schema
+        os_schema_path = schemas_dir / "endgame_ecs_mapping.json.gz"
+        compressed = gzip_compress(json.dumps(raw_os_schema, sort_keys=True, cls=DateTimeEncoder))
+        os_schema_path.write_bytes(compressed)
+        print(f"Endgame raw schema file saved: {os_schema_path}")
+
+
+class EndgameSchema(eql.Schema):
+    """Endgame schema for query validation."""
+
+    type_mapping = {
```

Comment:
if this is to be shared identically with its usage in ecs schemas, we should centralize it and share it for continuity sake

Reply:
actually looks like almost the entire object is a duplicate

Reply:
I thought about that, and it is similar, but not the same.

Reply:
I thought about that, and it is similar, but not the same. Most of the logic is already centralized per se in eql.Schema. We could try to get them closer if need be but it seems appropriate to leave as is.

Reply:
we can leave it as is for now for this PR, but we should look at all 3 schemas generation code to see about simplification

Reply:
https://github.com/elastic/detection-rules/issues/2370